### PR TITLE
Use anyNA before expensive is.na

### PR DESCRIPTION
**Pull request by HughParsonage**
Fairly minimal change with fairly dramatic performance improvements. Execution time of `as_factor(x)` for a 200,000 × 44 data.frame fell from 8 s to 1 s.

`anyNA` was introduced in R 3.1.0 so no change to dependencies.

### Benchmark

``` r
# The 2006 Australian Census 1% sample file
Census <- haven::read_dta("~/Data/ABS/Census2006_Stata/CSF06BP.dta")
prettyNum(dim(Census), big.mark = ",")
#> [1] "199,406" "47"

library(microbenchmark)
templib <- tempfile()
dir.create(templib)
install.packages("haven", lib = templib, quiet = TRUE)
#> package 'haven' successfully unpacked and MD5 sums checked
library("haven", lib.loc = templib)
microbenchmark(as_factor(Census), unit = "ms", times = 2L)
#> Unit: milliseconds
#>               expr      min       lq    mean  median       uq      max
#>  as_factor(Census) 7747.571 7747.571 7773.73 7773.73 7799.888 7799.888
#>  neval
#>      2
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.5.1 (2018-07-02)
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  English_Australia.1252      
#>  tz       Australia/Sydney            
#>  date     2018-08-29
#> Packages -----------------------------------------------------------------
#>  package        * version date       source        
#>  backports        1.1.2   2017-12-13 CRAN (R 3.5.0)
#>  base           * 3.5.1   2018-07-02 local         
#>  codetools        0.2-15  2016-10-05 CRAN (R 3.5.1)
#>  compiler         3.5.1   2018-07-02 local         
#>  crayon           1.3.4   2017-09-16 CRAN (R 3.5.1)
#>  datasets       * 3.5.1   2018-07-02 local         
#>  devtools         1.13.6  2018-06-27 CRAN (R 3.5.1)
#>  digest           0.6.16  2018-08-22 CRAN (R 3.5.1)
#>  evaluate         0.11    2018-07-17 CRAN (R 3.5.1)
#>  forcats          0.3.0   2018-02-19 CRAN (R 3.5.1)
#>  graphics       * 3.5.1   2018-07-02 local         
#>  grDevices      * 3.5.1   2018-07-02 local         
#>  grid             3.5.1   2018-07-02 local         
#>  haven          * 1.1.2   2018-06-27 CRAN (R 3.5.1)
#>  hms              0.4.2   2018-03-10 CRAN (R 3.5.1)
#>  htmltools        0.3.6   2017-04-28 CRAN (R 3.5.1)
#>  knitr            1.20    2018-02-20 CRAN (R 3.5.1)
#>  lattice          0.20-35 2017-03-25 CRAN (R 3.5.1)
#>  magrittr         1.5     2014-11-22 CRAN (R 3.5.1)
#>  MASS             7.3-50  2018-04-30 CRAN (R 3.5.1)
#>  Matrix           1.2-14  2018-04-13 CRAN (R 3.5.1)
#>  memoise          1.1.0   2017-04-21 CRAN (R 3.5.1)
#>  methods        * 3.5.1   2018-07-02 local         
#>  microbenchmark * 1.4-4   2018-01-24 CRAN (R 3.5.1)
#>  multcomp         1.4-8   2017-11-08 CRAN (R 3.5.1)
#>  mvtnorm          1.0-8   2018-05-31 CRAN (R 3.5.0)
#>  pillar           1.3.0   2018-07-14 CRAN (R 3.5.1)
#>  pkgconfig        2.0.2   2018-08-16 CRAN (R 3.5.1)
#>  R6               2.2.2   2017-06-17 CRAN (R 3.5.1)
#>  Rcpp             0.12.18 2018-07-23 CRAN (R 3.5.1)
#>  readr            1.1.1   2017-05-16 CRAN (R 3.5.1)
#>  rlang            0.2.2   2018-08-16 CRAN (R 3.5.1)
#>  rmarkdown        1.10    2018-06-11 CRAN (R 3.5.1)
#>  rprojroot        1.3-2   2018-01-03 CRAN (R 3.5.1)
#>  sandwich         2.5-0   2018-08-17 CRAN (R 3.5.1)
#>  splines          3.5.1   2018-07-02 local         
#>  stats          * 3.5.1   2018-07-02 local         
#>  stringi          1.1.7   2018-03-12 CRAN (R 3.5.0)
#>  stringr          1.3.1   2018-05-10 CRAN (R 3.5.1)
#>  survival         2.42-6  2018-07-13 CRAN (R 3.5.1)
#>  TH.data          1.0-9   2018-07-10 CRAN (R 3.5.1)
#>  tibble           1.4.2   2018-01-22 CRAN (R 3.5.1)
#>  tools            3.5.1   2018-07-02 local         
#>  utils          * 3.5.1   2018-07-02 local         
#>  withr            2.1.2   2018-03-15 CRAN (R 3.5.1)
#>  yaml             2.2.0   2018-07-25 CRAN (R 3.5.1)
#>  zoo              1.8-3   2018-07-16 CRAN (R 3.5.1)

detach("package:haven", unload = TRUE)

print(system("git log -1", intern = TRUE))
#> commit 09a773f16bfe7be108fda8b1c2c01ef1c1385a4f
#> Author: HughParsonage 
#> Date:   Fri Aug 24 21:09:35 2018 +1000

#>    anyNA may be redundant in this context

library(haven)
microbenchmark(as_factor(Census), unit = "ms", times = 2L)
#> Unit: milliseconds
#>               expr      min       lq     mean   median      uq     max
#>  as_factor(Census) 772.7193 772.7193 828.0886 828.0886 883.458 883.458
#>  neval
#>      2
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.5.1 (2018-07-02)
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  English_Australia.1252      
#>  tz       Australia/Sydney            
#>  date     2018-08-29
#> Packages -----------------------------------------------------------------
#>  package        * version    date       source        
#>  backports        1.1.2      2017-12-13 CRAN (R 3.5.0)
#>  base           * 3.5.1      2018-07-02 local         
#>  codetools        0.2-15     2016-10-05 CRAN (R 3.5.1)
#>  compiler         3.5.1      2018-07-02 local         
#>  crayon           1.3.4      2017-09-16 CRAN (R 3.5.1)
#>  datasets       * 3.5.1      2018-07-02 local         
#>  devtools         1.13.6     2018-06-27 CRAN (R 3.5.1)
#>  digest           0.6.16     2018-08-22 CRAN (R 3.5.1)
#>  evaluate         0.11       2018-07-17 CRAN (R 3.5.1)
#>  forcats          0.3.0      2018-02-19 CRAN (R 3.5.1)
#>  graphics       * 3.5.1      2018-07-02 local         
#>  grDevices      * 3.5.1      2018-07-02 local         
#>  grid             3.5.1      2018-07-02 local         
#>  haven          * 1.1.2.9000 2018-08-24 local         
#>  hms              0.4.2      2018-03-10 CRAN (R 3.5.1)
#>  htmltools        0.3.6      2017-04-28 CRAN (R 3.5.1)
#>  knitr            1.20       2018-02-20 CRAN (R 3.5.1)
#>  lattice          0.20-35    2017-03-25 CRAN (R 3.5.1)
#>  magrittr         1.5        2014-11-22 CRAN (R 3.5.1)
#>  MASS             7.3-50     2018-04-30 CRAN (R 3.5.1)
#>  Matrix           1.2-14     2018-04-13 CRAN (R 3.5.1)
#>  memoise          1.1.0      2017-04-21 CRAN (R 3.5.1)
#>  methods        * 3.5.1      2018-07-02 local         
#>  microbenchmark * 1.4-4      2018-01-24 CRAN (R 3.5.1)
#>  multcomp         1.4-8      2017-11-08 CRAN (R 3.5.1)
#>  mvtnorm          1.0-8      2018-05-31 CRAN (R 3.5.0)
#>  pillar           1.3.0      2018-07-14 CRAN (R 3.5.1)
#>  pkgconfig        2.0.2      2018-08-16 CRAN (R 3.5.1)
#>  R6               2.2.2      2017-06-17 CRAN (R 3.5.1)
#>  Rcpp             0.12.18    2018-07-23 CRAN (R 3.5.1)
#>  readr            1.1.1      2017-05-16 CRAN (R 3.5.1)
#>  rlang            0.2.2      2018-08-16 CRAN (R 3.5.1)
#>  rmarkdown        1.10       2018-06-11 CRAN (R 3.5.1)
#>  rprojroot        1.3-2      2018-01-03 CRAN (R 3.5.1)
#>  rstudioapi       0.7        2017-09-07 CRAN (R 3.5.1)
#>  sandwich         2.5-0      2018-08-17 CRAN (R 3.5.1)
#>  splines          3.5.1      2018-07-02 local         
#>  stats          * 3.5.1      2018-07-02 local         
#>  stringi          1.1.7      2018-03-12 CRAN (R 3.5.0)
#>  stringr          1.3.1      2018-05-10 CRAN (R 3.5.1)
#>  survival         2.42-6     2018-07-13 CRAN (R 3.5.1)
#>  TH.data          1.0-9      2018-07-10 CRAN (R 3.5.1)
#>  tibble           1.4.2      2018-01-22 CRAN (R 3.5.1)
#>  tools            3.5.1      2018-07-02 local         
#>  utils          * 3.5.1      2018-07-02 local         
#>  withr            2.1.2      2018-03-15 CRAN (R 3.5.1)
#>  yaml             2.2.0      2018-07-25 CRAN (R 3.5.1)
#>  zoo              1.8-3      2018-07-16 CRAN (R 3.5.1)
```

Created on 2018-08-29 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # haven 1.1.2.9000
 
+* `as_factor(x)` is now faster when `x` does not contain any missing values
```
Comment (hadley) [inline NEWS.md]: Please acknowledge yourself as described in http://style.tidyverse.org/news.html#acknowledgement

File: R/as_factor.R [hunk 2]
```diff
@@ -98,7 +102,11 @@ replace_with <- function(x, from, to) {
   }
 
   matches <- match(na_tag(x), na_tag(from), incomparables = NA)
-  out[!is.na(matches)] <- to[matches[!is.na(matches)]]
-
+  #
+  # if (anyNA(matches)) {
```
[the remark this answers is not in the document]
Comment (HughParsonage) [reply R/as_factor.R]: I did. I couldn't work out whether the commented out code is redundant is not. Is `anyNA(matches)` always `TRUE` here? I couldn't create tests that enter the else branch.

File: NEWS.md (hunk 1 above)
Comment (HughParsonage) [reply NEWS.md]: Done.

File: R/as_factor.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (hadley) [reply R/as_factor.R]: I'm confused. The branch wasn't needed before, so why would we need it now?

**Comment by HughParsonage** [comment]
Performance. But only if anyNA(matches) can be FALSE.

I didn’t identify it as a bottleneck but as it was exactly the same as the
code that did form the bottleneck, I made the same change here. I got a
codecov failure when I added it however. And since I couldn’t construct a
test that would cover both cases, I commented it out and flagged it in the
hope that a reviewer with more of an understanding of this section could
quickly see how to test it or that it would be futile to test.

Sorry if it’s sucked up too much time: we can just remove the comments. (On
my phone atm so can’t push for about 12 hrs.)

**Comment by hadley** [comment]
Yeah, let's leave it for now.

**Comment by hadley** [comment]
Thanks!

**Pull request merged**